rfcs/1-input-pipeline.md

```diff
@@ -0,0 +1,391 @@
+                )
+        return X, y, sample_weight
+    
+    def get_meta(self) -> Dict[str, Any]:
+        return {"n_features_in_": self.n_feautres_in_}
+
+
+class ValidateClassifierTargetArray(ArrayTransformer):
+
+    def transform_input(self, X: ArrayLike, y: Optional[ArrayLike], sample_weight: Optional[ArrayLike], *, initialize: bool) -> Data:
+        if y is not None:
+            y = check_array(
+                y,
+                ensure_2d=False,
+                allow_nd=False,
+                dtype=_check_array_dtype(y, force_numeric=False),
+            )
+            classes_ = np.unique(y)
+            if initialize:
+                self.classes_ = classes_
+            else:
+                if self.classes_ != classes_:
+                    raise ValueError(
+                        f"Expected y to have {self.classes_} classes, but got {classes_} classes"
+                    )
+        return X, y, sample_weight
+
+    def get_meta(self) -> Dict[str, Any]:
+        return {"classes_": self.classes_}
+
+
+class ValidateRegressorTargetArray(ArrayTransformer):
+
+    def transform_input(self, X: ArrayLike, y: Optional[ArrayLike], sample_weight: Optional[ArrayLike], *, initialize: bool) -> Data:
+        if y is not None:
+            y = check_array(
+                y,
+                ensure_2d=False,
+                allow_nd=False,
+                dtype=_check_array_dtype(y, force_numeric=True),
+            )
+        return X, y, sample_weight
+
+
+class ValidateSampleWeight(ArrayTransformer):
+
+    def transform_input(self, X: ArrayLike, y: Optional[ArrayLike], sample_weight: Optional[ArrayLike], *, initialize: bool) -> Data:
+        if isinstance(sample_weight, numbers.Number):
+            sample_weight = np.full(shape=(len(X),), fill_value=sample_weight)
+        if sample_weight is not None:
+            sample_weight = check_array(
+                sample_weight,
+                accept_sparse=False,
+                ensure_2d=False,
+                dtype=tf.keras.backend.floatx(),
+                copy=False,
+            )
+            if sample_weight.ndim != 1:
+                raise ValueError("Sample weights must be 1D array or scalar")
+            if np.all(sample_weight == 0):
+                raise ValueError(
+                    "No training samples had any weight; only zeros were passed in sample_weight."
+                    " That means there's nothing to train on by definition, so training can not be completed."
+                )
+        return X, y, sample_weight
+
+```
+
+Now we can add this to the default array transformers:
+
+```python
+class KerasClassifier:
+    def __init__(
+        self,
+        array_transformers = (ClassifierInputValidator,)
+    ):
+        ...
+
+
+class KerasRegressor:
+    def __init__(
+        self,
+        array_transformers = (RegressorInputValidator,)
+    ):
+        ...
+```
+
+## Issues this can potentially resolve
```

Comment:
These issues seem to be around modular validation/transforming. This RFC proposes one solution. What are other solutions? Why does this RFC represent the best solution?

Reply:
> What are other solutions?

This is, of course, a great question. I think general solutions would be variations of the same idea, perhaps less structured (eg. hardcoding that `tf.data.Dataset` inputs should skip `BaseWrapper._validate_data`). I'd have to think a bit to see if I can come up with any other structured approaches that might also solve the issues.

Reply:
Why not have two Scikit-learn transformers, one for validation and one to change the data?

Reply:
We would need:
1. Interface for validation using array-like data
2. Interface for data changing using array-like data
3. Interface for validation using Dataset
4. Interface for data changing using array-like data

I think this is too many interfaces. The only difference between a "validation" and a "transformation" is that the transformation _needs_ to return the data, but the validation does not. So by having the validation return the data we can collapse those two concepts into one. Validation transformers simply inspect the data but do not modify it.

Reply:
What use case/issue motivates having separate classes for Dataset/array-like inputs? Why not collapse it into one class?

Why is 4 interfaces "too many"? Why is the current framework with `target_encoder_` and `feature_encoder_` not sufficient?

I'm asking questions to get answers encoded in the RFC (an RFC is a really good idea to encode these decisions).